### PR TITLE
Integrate PPR for pathfinding between stop places

### PR DESCRIPTION
**Pull request by janikkaden**
The [Per Pedes Routing](https://motis-project.de/docs/api/endpoint/ppr.html) submodule from the [MOTIS project](http://motis-project.de/) is used to generate paths between stop places. 

- Everything runs inside docker containers.
- The containers are started from an docker compose file.
- A python script gets all stop places from the PGSQL database and feeds them as start/stop positions to the PPR API.

### CODE REVIEW

**Review by Robbendebiene — COMMENTED** [review]
We should be able to remove `highways.lua` (make sure to remove any references from `main.lua` as well)

File: scripts/ppr.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        for ii in range(i + 1 , len(stop_places)):
+            
+            json_data = makeRequest(url, payload,stop_places,i,ii)
+            insertPGSQL(cur,json_data,stop_places,path_counter,i,ii)
+            path_counter = path_counter + 1
+            
+            # generate bidirectional paths:
+            json_data = makeRequest(url, payload,stop_places,ii,i)
+            insertPGSQL(cur,json_data,stop_places,path_counter,ii,i)
+            path_counter = path_counter + 1
```
Comment (Robbendebiene) [inline scripts/ppr.py:121]: Since this will double the amount of requests we will have to make to PPR it would be nice if we could just invert the result we got above. I also fear it might give a different reverse way in some special scenarios, where I'm not sure if this is good or bad. If you have any thoughts on this please let me know.

Inverting should be rather trivial, since we just need to swap the DHIDs and maybe for extra safety we also reverse the geometry using https://postgis.net/docs/ST_Reverse.html (no additional call to `insertPathsElementsRef` needed)

File: scripts/ppr.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    # PPR can return multiple possible paths for one connection:
+    for route in json_data["routes"]:
+        path = route["path"]
+        edges = route["edges"]
+        # distance = route["distance"]
+        relation_id = stop_places[ii]["relation_id"]
+                
+        insertPathsElementsRef(cur, edges, path_counter)
+        insertPath(cur, relation_id, stop_places[i]["IFOPT"], stop_places[ii]["IFOPT"], path)
+
```
Comment (Robbendebiene) [inline scripts/ppr.py:55]: Just some minor coding style suggestion.
Instead of passing the array with indexes it seems more common to me to directly pass the actual values/contents of the array at these indexes. But maybe you have a good reason for not doing so.

File: scripts/ppr.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    except Exception as e:
+        conn.close()
+        exit(e)
+
+    path_counter = 1
+    
+    print("retrieving paths ...")
+
+    # Iterate through all stop_places to get the path between all of them.
```
Comment (Robbendebiene) [inline scripts/ppr.py:110]: Won't this calculate the paths between every stop/quay no matter if they belong to the same stop area or not?

File: scripts/ppr.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+def main():
+    # Connect to PostgreSQL database
+    conn = psycopg2.connect(
+        host = os.environ['host_postgis'],
+        port = "5432",
+        database = "osm2vdv462",
+        user = "admin",
+        password = "admin"
+    )
```
[the remark this answers is not in the document]
Comment (janikkaden) [reply scripts/ppr.py]: It is included in commit f5bc86e

File: scripts/ppr.py (hunk 1 above)
Comment (janikkaden) [reply scripts/ppr.py:121]: > I also fear it might give a different reverse way in some special scenarios, where I'm not sure if this is good or bad.
This was the reason I made the requests in boths directions. If there may be different paths for the both directions, we would not able to depict that. But the only scenarios I can imagine where this could be relevant are staircases with an escalator only in one direction and airports with unidirectional gates/doors.

It is questionable if these special cases are relevant enough to let PPR calculate twice the amount of paths.

File: docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+networks:
+  osm2vdv462_net:
+    name: ${DOCKER_NETWORK}
+
+# Create docker volume which is a storage point located outside of containers.
+# This is required to persistently store the database between docker restarts.
+volumes:
+  postgis:
+  ppr:
+  pgadmin:
```
Comment (Robbendebiene) [inline docker-compose.yml:12]: Just for clarification. Why did you add an additional volume for pgadmin?

File: scripts/ppr.py (hunk 1 above)
Comment (Robbendebiene) [reply scripts/ppr.py:121]: For now I would stick to your implementation, but adding additional comments in the code describing the special cases this double computation" is catching.

File: docker-compose.yml (hunk 5 above)
Comment (janikkaden) [reply docker-compose.yml:12]: The Pgadmin container creates a volume by itself to permanently save contents of the folder "/var/lib/pgadmin". By creating the volume with a defined name and linking it to "pgadmin:/var/lib/pgadmin" you can get around it being named randomly by docker.
You then have three consistently named volumes (osm2vdv462_pgadmin, osm2vdv462_postgis, osm2vdv462_ppr).

File: scripts/ppr.py (hunk 3 above)
Comment (janikkaden) [reply scripts/ppr.py:110]: Yes you are right. I will update the code to seperately handle the stop areas.

File: scripts/ppr.py (hunk 2 above)
Comment (janikkaden) [reply scripts/ppr.py]: Thank you for the suggestion, i have changed this section in commit 85c5f97.

File: scripts/ppr.py (hunk 3 above)
Comment (janikkaden) [reply scripts/ppr.py]: Now fixed in commit 529192d

File: scripts/pgsql/stop_places.sql [hunk 6]
```diff
@@ -179,7 +143,7 @@ BEGIN
   ELSEIF tags->>'subway' = 'yes'
     THEN result := 'metroPlatform';
   ELSEIF tags->>'tram' = 'yes' THEN
-    IF ST_GeometryType(geom) = 'ST_Point'
+    IF geom::geometry = 'ST_Point'
```
Comment (Robbendebiene) [inline scripts/pgsql/stop_places.sql:146]: I feel like this will check if the content of `geom` is equal to `'ST_Point'` and not the type of `geom` as before.

Comment (janikkaden) [reply scripts/pgsql/stop_places.sql]: Thanks for the notice. I have changed it to `ST_GeometryType(geom::geometry)` in commit 603297a

File: scripts/ppr.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+def insertPathsElementsRef(cur, edges, path_counter):
+    osm_way_ids = []
+    for edge in edges:
+        osm_way_id = abs(edge["osm_way_id"])
+        # currently also the crossed ways are included (negative way osm id)
+        if osm_way_id != 0 and osm_way_id not in osm_way_ids:
+            osm_way_ids.append(osm_way_id)
+            cur.execute(
+                "INSERT INTO paths_elements_ref (path_id, osm_type, osm_id) VALUES (%s, %s, %s)",
+                (path_counter, 'W', osm_way_id)
```
[the remark this answers is not in the document]
Comment (janikkaden) [reply scripts/ppr.py:24]: PPR is not explicitly returning the osm type. For example an edge using an elevator is returned as following:
```
{
    "edge_type": "elevator",
    "street_type": "none",
    "crossing_type": "generated",
    "side": "center",
    "osm_way_id": -9121043127,
    "name": "",
    "marked_crossing_detour": 0,
    "distance": 0.0,
    ...
},
```

These are the edge types that PPR uses:

>     CONNECTION
>     STREET
>     FOOTWAY
>     CROSSING
>     ELEVATOR

Crossings always return the way that is crossed, even if there is a crossing node, so I think that elevator will be the only type that uses nodes. It may be possible to distinguish between ways and nodes in this way. But this also means, that the tags from crossing nodes get lost (information regarding traffic_signals, ...).

File: scripts/ppr.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        conn.close()
+        exit(e)
+    
+    final_stop_places = []
+
+    try:
+        # get all relevant stop places
+        cur.execute(
+            'SELECT relation_id, "IFOPT" FROM final_stop_places'
+        )
```
Comment (Robbendebiene) [inline scripts/ppr.py:97]: obsolete, query `stop_area_elements` table once (see other comment below)

**Review by Robbendebiene — CHANGES_REQUESTED** [review]

File: scripts/pgsql/stop_places.sql [hunk 9]
```diff
@@ -510,7 +475,8 @@ CREATE OR REPLACE TEMPORARY VIEW stop_places_with_organisations AS (
  * Aggregate member stop geometries to stop areas
  * Split JOINs because GROUP BY doesn't allow grouping by all columns of a specific table
  */
-CREATE OR REPLACE TEMPORARY VIEW final_stop_places AS (
+DROP TABLE IF EXISTS final_stop_places CASCADE;
+CREATE TABLE final_stop_places AS (
```
[the remark this answers is not in the document]
Comment (janikkaden) [reply scripts/pgsql/stop_places.sql]: Changed in e092ff5

File: README.md [hunk 10]
```diff
@@ -4,26 +4,30 @@ This repository offers a shell script to convert OpenStreetMap data (public tran
...
-The export pipeline requires [docker](https://www.docker.com/) and [osm2pgsql](https://osm2pgsql.org/) to be installed.
-You can find the respective installation instructions here:
-- [How to install docker](https://docs.docker.com/engine/install/)
-- [How to install osm2pgsql](https://osm2pgsql.org/doc/install.html)
+The export pipeline requires [docker](https://www.docker.com/) to be installed.
+You can find the respective installation instruction here: [how to install docker](https://docs.docker.com/engine/install/).
+
+The OSM data to be used can be downloaded from [geofabrik](https://download.geofabrik.de/). There you can find OSM extracts for all sorts of countries and regions. It has to be in the *.osm.pbf format. Please have in mind that there are high memory and RAM requirements for bigger regions as e.g. germany. On smaller machines you should use smaller extracts for testing.
+
+For generating the paths between stop places, the [Per Pedes Routing](https://motis-project.de/docs/api/endpoint/ppr.html) submodule from the [MOTIS project](http://motis-project.de/) is used.
```
[the remark this answers is not in the document]
Comment (janikkaden) [reply README.md]: Changed in e092ff5

File: scripts/ppr.py [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+    except Exception as e:
+        exit(e)
+
+    if response.status_code != 200:
+        exit(f'Request failed with status code {response.status_code}: {response.text}')
+
+    return response.json()
+
+
+def insertPGSQL(cur,json_data,start, stop ,path_counter):
```
[the remark this answers is not in the document]
Comment (janikkaden) [reply scripts/ppr.py]: Improved in 81276d0

File: scripts/ppr.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+
+    path_counter = 1
+
+    # Iterate through all stop places to get the path between the stops/quays.
+    for stop_place in final_stop_places:
+        stop_places = []
+        
+        try:
+            # SRID in POSTGIS is default set to 4326 --> x = lng, Y = lat
+            cur.execute(
```
[the remark this answers is not in the document]
Comment (janikkaden) [reply scripts/ppr.py]: Improved in 81276d0

**Review by Robbendebiene — APPROVED** [review]
LGTM!
